pkg/ddevapp/providerDrudS3.go

```diff
@@ -0,0 +1,426 @@
+// It returns an array of environment names and err
+func (p *DrudS3Provider) findDrudS3Project() (map[string]interface{}, error) {
+	_, client, err := p.getDrudS3Session()
+	if err != nil {
+		return nil, err
+	}
+	// Get a list of all projects the current user has access to.
+	projectMap, err := getDrudS3Projects(client, p.S3Bucket)
+	if err != nil {
+		return nil, err
+	}
+
+	projectWithEnvs := projectMap[p.app.Name]
+
+	// Get a list of environments for a given project.
+	return projectWithEnvs, nil
+}
+
+// getDrudS3Session loads the DrudS3 API config from disk and returns a DrudS3 session struct.
+func (p *DrudS3Provider) getDrudS3Session() (*session.Session, *s3.S3, error) {
+	sess, err := session.NewSession(&aws.Config{
+		Region:      aws.String("us-east-1"),
+		Credentials: credentials.NewStaticCredentials(p.AWSAccessKey, p.AWSSecretKey, ""),
+	},
+	)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to NewSession: %v", err)
+	}
+
+	client := s3.New(sess)
+
+	return sess, client, nil
+}
+
+// Functions to allow golang sort-by-modified-date (descending) for s3 objects found.
+type byModified []*s3.Object
+
+func (objs byModified) Len() int {
+	return len(objs)
+}
+func (objs byModified) Swap(i, j int) {
+	objs[i], objs[j] = objs[j], objs[i]
+}
+func (objs byModified) Less(i, j int) bool {
+	return objs[j].LastModified.Before(*objs[i].LastModified)
+}
+
+// getDrudS3Projects returns a map of project[environment] so it's easy to
+// find what projects are available in the bucket and then to get the
+// environments for that project.
+func getDrudS3Projects(client *s3.S3, bucket string) (map[string]map[string]interface{}, error) {
+	objects, err := getS3ObjectsWithPrefix(client, bucket, "")
+	if err != nil {
+		return nil, err
+	}
+	projectMap := make(map[string]map[string]interface{})
+
+	// This sadly is processing all of the items we receive, all the files in all the directories
+	for _, obj := range objects {
+		// TODO: It might be possible but unlikely for the object key separator not to be a "/"
+		components := strings.Split(strings.TrimRight(*obj.Key, "/"), "/")
+		// We're only interested in items that have an actual dump in them.
+		if (len(components)) == 3 {
+			if _, ok := projectMap[components[0]]; !ok {
+				tmp := make(map[string]interface{})
+				projectMap[components[0]] = tmp
+			}
+			projectMap[components[0]][components[1]] = true
+		}
+	}
+	return projectMap, nil
+}
+
+// getS3ObjectsWithPrefix gets all S3 objects in the named bucket with the prefix provided.
+func getS3ObjectsWithPrefix(client *s3.S3, bucket string, prefix string) ([]*s3.Object, error) {
+	// TODO: This may be fragile because it could return a lot of items.
+	maxKeys := aws.Int64(1000000000)
+
+	// TODO: WARNING: ListObjects only returns first 1000 objects
+	resp, err := client.ListObjects(&s3.ListObjectsInput{
+		Bucket:  aws.String(bucket),
+		Prefix:  aws.String(prefix),
+		MaxKeys: maxKeys,
+	})
```

Comment:
I'm ok with adding this, and of course you're correct. OTOH, the upstream project is supposed to prune to a few weeklies and a few dailies. Do you think it's worth going ahead and adding the (hard-to-test) loop?

Reply:
If it's truly never expected to come up, then that's totally fair. Given that assumption, though, I think the way MaxKeys is currently defined is a bit confusing, and there may be a reasonable middle ground where we just simply check if S3 indicates that more results are available, then warn or fail on that.

Reply:
I will do the loop. My fear was that I'd write code that didn't get executed enough and would be basically untested. But silly me, I realized I can make the max *low* and it will be exercised regularly. Or at least during testing.